File: clinical/src/main/java/com/hartwig/actin/clinical/feed/lab/LabUnitResolver.java
package com.hartwig.actin.clinical.feed.lab;

import java.util.Map;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Maps;
import com.hartwig.actin.clinical.datamodel.LabUnit;

import org.jetbrains.annotations.NotNull;

final class LabUnitResolver {

    @VisibleForTesting
    static final Map<String, LabUnit> CURATION_MAP = Maps.newHashMap();

    static {
        CURATION_MAP.put("10*6/L", LabUnit.MILLIONS_PER_LITER);
        CURATION_MAP.put("10*9/L", LabUnit.BILLIONS_PER_LITER);
        CURATION_MAP.put("10*12/L", LabUnit.TRILLIONS_PER_LITER);
        CURATION_MAP.put("% van de leukocyten", LabUnit.PERCENTAGE_OF_LEUKOCYTES);
        CURATION_MAP.put("% van de T-cellen", LabUnit.PERCENTAGE_OF_T_CELLS);

        CURATION_MAP.put("mmol/mol Kreatinine", LabUnit.MILLIMOLES_PER_MOLE);
        CURATION_MAP.put("µg/L", LabUnit.MICROGRAMS_PER_LITER);
        CURATION_MAP.put("µmol/L", LabUnit.MICROMOLES_PER_LITER);
        CURATION_MAP.put("micromol/l", LabUnit.MICROMOLES_PER_LITER);
        CURATION_MAP.put("E/ml", LabUnit.UNITS_PER_MILLILITER);

        // L/L is an implied unit used for hematocrit
        CURATION_MAP.put("L/L", LabUnit.NONE);

        CURATION_MAP.put("mol/mol", LabUnit.NONE);
    }

    private LabUnitResolver() {
    }

    @NotNull
    public static LabUnit resolve(@NotNull String unit) {
        if (CURATION_MAP.containsKey(unit)) {
            return CURATION_MAP.get(unit);
        }

        for (LabUnit labUnit : LabUnit.values()) {
            if (labUnit.display().equalsIgnoreCase(unit)) {
                return labUnit;
            }
        }

        throw new IllegalStateException("Could not map lab unit: '" + unit + "'");
    }
}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/lab/LabUnitResolver.kt
package com.hartwig.actin.clinical.feed.emc.lab

import com.hartwig.actin.datamodel.clinical.LabUnit

object LabUnitResolver {

    val CURATION_MAP = mapOf(
        "10*6/L" to LabUnit.MILLIONS_PER_LITER,
        "*10E6/l" to LabUnit.MILLIONS_PER_LITER,
        "10*9/L" to LabUnit.BILLIONS_PER_LITER,
        "*10E9/l" to LabUnit.BILLIONS_PER_LITER,
        "10*12/L" to LabUnit.TRILLIONS_PER_LITER,
        "% van de leukocyten" to LabUnit.PERCENTAGE_OF_LEUKOCYTES,
        "% van de T-cellen" to LabUnit.PERCENTAGE_OF_T_CELLS,
        "mmol/mol Kreatinine" to LabUnit.MILLIMOLES_PER_MOLE,
        "g/mol Kreatinine" to LabUnit.GRAMS_PER_MOLE,
        "µg/L" to LabUnit.MICROGRAMS_PER_LITER,
        "µmol/L" to LabUnit.MICROMOLES_PER_LITER,
        "micromol/l" to LabUnit.MICROMOLES_PER_LITER,
        "E/ml" to LabUnit.UNITS_PER_MILLILITER,
        // L/L is an implied unit used for hematocrit
        "L/L" to LabUnit.NONE,
        "mol/mol" to LabUnit.NONE,
        "Ratio" to LabUnit.NONE
    )

    fun resolve(unit: String) = CURATION_MAP[unit]
        ?: LabUnit.entries.firstOrNull { it.display().equals(unit, ignoreCase = true) }
        ?: throw IllegalStateException("Could not map lab unit: '$unit'")
}